devito/tools/data_structures.py

```diff
@@ -480,3 +481,75 @@ def __hash__(self):
                 h ^= hash((key, value))
             self._hash = h
         return self._hash
+
+
+class UnboundedMultiTuple(object):
+
+    """
+    An UnboundedMultiTuple is an ordered collection of tuples that can be
+    infinitely iterated over.
+
+    Examples
+    --------
+    >>> ub = UnboundedMultiTuple([1, 2], [3, 4])
+    >>> ub
+    UnboundedMultiTuple((1, 2), (3, 4))
+    >>> ub.iter()
+    >>> ub
+    UnboundedMultiTuple((1, 2)*, (3, 4))
+    >>> ub.next()
+    1
+    >>> ub.next()
+    2
+    >>> ub.next()
+    2
+    >>> ub.iter()
+    >>> ub.iter()  # No effect, tip has reached the last tuple
+    >>> ub.iter()  # No effect, tip has reached the last tuple
+    >>> ub
+    UnboundedMultiTuple((1, 2), (3, 4)*)
+    >>> ub.next()
+    3
+    >>> ub.next()
+    4
+    >>> ub.iter()  # Reloads the last iterator
+    >>> ub.next()
+    3
+    """
+
+    def __init__(self, *items):
+        # Normalize input
+        nitems = []
+        for i in as_tuple(items):
+            if isinstance(i, Iterable):
+                nitems.append(tuple(i))
+            else:
+                raise ValueError("Expected sequence, got %s" % type(i))
+
+        self.items = tuple(nitems)
+        self.tip = -1
+        self.curiter = None
+
+    def __repr__(self):
+        items = []
```

Comment:
```
items = list(self.nitems)
insert(items, "*", self.tip-1)
```

Reply:
you're right, improved as per your suggestion